src/execution_plan/execution_plan.c

```diff
@@ -8,695 +8,804 @@
+    // Handle argument entities
+    uint arg_count = cypher_ast_call_narguments(call_clause);
+    // if (expressions == NULL) expressions = array_new(AR_ExpNode*, arg_count);
+    const char **arguments = array_new(const char*, arg_count);
+    for (uint i = 0; i < arg_count; i ++) {
 
-    if(root->type == type) {
-        return root;
-    }
+        const cypher_astnode_t *ast_exp = cypher_ast_call_get_argument(call_clause, i);
 
-    for(int i = 0; i < root->childCount; i++) {
-        OpBase *op = ExecutionPlan_LocateOp(root->children[i], type);
-        if(op) return op;
+        const cypher_astnode_t *identifier_node = cypher_ast_projection_get_alias(ast_exp);
+        const char *identifier = cypher_ast_identifier_get_name(identifier_node);
+
+        arguments = array_append(arguments, identifier);
     }
 
-    return NULL;
+    return arguments;
 }
 
-void _ExecutionPlan_LocateOps(OpBase *root, OPType type, OpBase ***ops) {
-    if(!root) return;
+static void _ExecutionPlanSegment_ProcessQueryGraph(ExecutionPlanSegment *segment, QueryGraph *qg, FT_FilterNode *ft, rax *resolved, Vector *ops) {
+    GraphContext *gc = GraphContext_GetFromTLS();
+    AST *ast = AST_GetFromTLS();
 
-    if(root->type & type) (*ops) = array_append((*ops), root);
+    QueryGraph **connectedComponents = QueryGraph_ConnectedComponents(qg);
+    uint connectedComponentsCount = array_len(connectedComponents);
+    segment->connected_components = connectedComponents;
 
-    for(int i = 0; i < root->childCount; i++) {
-        _ExecutionPlan_LocateOps(root->children[i], type, ops);
+    /* For every connected component.
+     * Incase we're dealing with multiple components
+     * we'll simply join them all together with a join operation. */
+    OpBase *cartesianProduct = NULL;
+    if(connectedComponentsCount > 1) {
+        cartesianProduct = NewCartesianProductOp();
+        Vector_Push(ops, cartesianProduct);
     }
-}
 
-OpBase** ExecutionPlan_LocateOps(OpBase *root, OPType type) {
-    OpBase **ops = array_new(OpBase*, 0);
-    _ExecutionPlan_LocateOps(root, type, &ops);
-    return ops;
-}
+    // Keep track after all traversal operations along a pattern.
+    Vector *traversals = NewVector(OpBase*, 1);
+    OpBase *op;
 
-void ExecutionPlan_Taps(OpBase *root, OpBase ***taps) {
-    if(root == NULL) return;
-    if(root->type & OP_SCAN) *taps = array_append(*taps, root);
+    for(uint i = 0; i < connectedComponentsCount; i++) {
+        QueryGraph *cc = connectedComponents[i];
+        uint edge_count = array_len(cc->edges);
+        if(edge_count == 0) {
+            /* Node scan. */
+            QGNode *n = cc->nodes[0];
+            uint rec_idx = RecordMap_FindOrAddID(segment->record_map, n->id);
+            if(n->labelID != GRAPH_NO_LABEL) op = NewNodeByLabelScanOp(n, rec_idx);
+            else op = NewAllNodeScanOp(gc->g, n, rec_idx);
+            Vector_Push(traversals, op);
+            _UpdateResolvedVariables(resolved, op);
```

Comment:
This might be dangerous depending to what extend resolved is used,
as in the optimisation phase op might be removed / deleted.

Reply:
I don't think `resolved` actually does _anything_ right now; it seems like it can be deleted entirely.

Reply:
Then let's remove it